### PR TITLE
fix(optimizer): fix filter pushdown and range fold

### PR DESCRIPTION
**Pull request by kysshsy**
related  to #794 #834 , add some new rules and change scan's default filter to true(this make it match with the "filter-scan" rule)

### CODE REVIEW

File: src/planner/rules/expr.rs [hunk 1]
```diff
@@ -172,6 +184,47 @@ fn is_not_zero(var: &str) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
...
+    f: impl Fn(&DataValue, &DataValue) -> bool,
+) -> impl Fn(&mut EGraph, Id, &Subst) -> bool {
+    let v1 = var(v1);
+    let v2 = var(v2);
+
+    move |egraph, _, subst| match (
+        &egraph[subst[v1]].data.constant,
+        &egraph[subst[v2]].data.constant,
+    ) {
+        (Some(d1), Some(d2)) => f(d1, d2),
```
Comment (wangrunji0408) [inline src/planner/rules/expr.rs:223]: Seems like the common condition `std::mem::discriminant(d1) == std::mem::discriminant(d2)` can be moved to here.

**Review by wangrunji0408 — APPROVED** [review]
LGTM. Thank you!

Should we get #795 merged first? @skyzh

**Comment by skyzh** [comment]
yep i'll handle these two pull requests on the weekends :)

**Comment by skyzh** [comment]
need rebasing + fix conflicts, I can help with that next weekend :)

**Comment by kysshsy** [comment]
rebase the commit, help me check the sql test result, thank you 😁

**Pull request merged**